src/Agent.Worker/Release/Artifacts/BuildArtifact.cs

```diff
@@ -0,0 +1,189 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+using Agent.Worker.Release.Artifacts.Definition;
+
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.VisualStudio.Services.Agent;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.Agent.Worker;
+using Microsoft.VisualStudio.Services.Client;
+using Microsoft.VisualStudio.Services.FileContainer;
+
+using ServerBuildArtifact = Microsoft.TeamFoundation.Build.WebApi.BuildArtifact;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Release.Artifacts
+{
+    public class BuildArtifact : IArtifact
+    {
+        public const string AllArtifacts = "*";
+
+        public async Task Download(ArtifactDefinition artifactDefinition, IHostContext hostContext, IExecutionContext executionContext, string localFolderPath)
+        {
+            ArgUtil.NotNull(artifactDefinition, nameof(artifactDefinition));
+            ArgUtil.NotNull(hostContext, nameof(hostContext));
+            ArgUtil.NotNull(executionContext, nameof(executionContext));
+            ArgUtil.NotNullOrEmpty(localFolderPath, nameof(localFolderPath));
+
+            int buildId = Convert.ToInt32(artifactDefinition.Version, CultureInfo.InvariantCulture);
+            if (buildId <= 0)
+            {
+                throw new ArgumentException("artifactDefinition.Version");
+            }
+
+            var buildArtifactDetails = artifactDefinition.Details as BuildArtifactDetails;
+            if (buildArtifactDetails == null)
+            {
+                throw new ArgumentException("artifactDefinition.Details");
+            }
+
+            // Get the list of available artifacts from build. 
+            executionContext.Output(StringUtil.Loc("RMPreparingToGetBuildArtifactList"));
+
+            var vssConnection = new VssConnection(buildArtifactDetails.TfsUrl, buildArtifactDetails.Credentials);
+            var buildClient = vssConnection.GetClient<BuildHttpClient>();
+            var xamlBuildClient = vssConnection.GetClient<XamlBuildHttpClient>();
+            List<ServerBuildArtifact> buildArtifacts = null;
+            DefinitionType buildDefinitionType = DefinitionType.Build;
+            try
+            {
+                buildArtifacts = await buildClient.GetArtifactsAsync(buildArtifactDetails.Project, buildId);
+            }
+            catch (AggregateException ex)
```

Comment:
correct.. it does not give aggregateexception.. changed to BuildNotFoundException..